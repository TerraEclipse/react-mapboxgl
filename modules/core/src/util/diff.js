import _ from 'lodash'

/**
 * Produces a shallow diff of an object. Output is an array of
 * 'changes' where a change is either an 'add', 'remove', or 'update'.
 */
export default function diff (a, b) {
  let changes = []
  let keysA = _.keys(a)
  let keysB = _.keys(b)

  // Find the items in A that are not in B.
  _.each(_.difference(keysA, keysB), (key) => {
    changes.push({type: 'remove', key: key, value: a[key]})
  })

  // Find the items in B that are not in A.
  _.each(_.difference(keysB, keysA), (key) => {
    changes.push({type: 'add', key: key, value: b[key]})
  })

  // Find the items that are in both, but have changed.
  _.each(_.intersection(keysA, keysB), (key) => {
    if (!_.isEqual(a[key], b[key])) {
      changes.push({type: 'update', key: key, value: b[key]})
    }
  })

  return changes
}
