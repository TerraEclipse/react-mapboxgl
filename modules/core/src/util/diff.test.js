/* eslint-env jest */
import diff from './diff'

describe('diff', () => {
  it('can diff two objects', () => {
    var a = {
      meal: 'dinner',
      name: 'quiche',
      ingredients: [
        'eggs',
        'cream',
        'cheese'
      ],
      status: {
        prepared: true,
        baked: false,
        ate: false
      },
      favorite: true
    }
    var b = {
      meal: 'dinner',
      name: 'bacon quiche',
      ingredients: [
        'eggs',
        'cream',
        'cheese',
        'bacon'
      ],
      status: {
        prepared: true,
        baked: true,
        ate: true
      },
      vegetarian: false
    }
    var changes = diff(a, b)
    expect(changes).toMatchSnapshot()
  })
})
