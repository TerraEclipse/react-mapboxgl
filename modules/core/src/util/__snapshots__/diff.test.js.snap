// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff can diff two objects 1`] = `
Array [
  Object {
    "key": "favorite",
    "type": "remove",
    "value": true,
  },
  Object {
    "key": "vegetarian",
    "type": "add",
    "value": false,
  },
  Object {
    "key": "name",
    "type": "update",
    "value": "bacon quiche",
  },
  Object {
    "key": "ingredients",
    "type": "update",
    "value": Array [
      "eggs",
      "cream",
      "cheese",
      "bacon",
    ],
  },
  Object {
    "key": "status",
    "type": "update",
    "value": Object {
      "ate": true,
      "baked": true,
      "prepared": true,
    },
  },
]
`;
